adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleBanner.m

```diff
@@ -101,154 +98,133 @@ - (void)requestBannerAd {
+  NSArray<NSValue *> *potentials = @[
+    NSValueFromGADAdSize(GADAdSizeBanner), NSValueFromGADAdSize(GADAdSizeLeaderboard),
+    NSValueFromGADAdSize(shortBannerSize)
+  ];
+
+  GADAdSize closestSize = GADClosestValidSizeForAdSizes(adSize, potentials);
+  CGSize size = CGSizeFromGADAdSize(closestSize);
+  if (size.height == GADAdSizeBanner.size.height) {
+    if (size.width < GADAdSizeBanner.size.width) {
+      return shortBannerSize;
+    } else {
+      return GADAdSizeBanner;
     }
+  } else if (size.height == GADAdSizeLeaderboard.size.height) {
+    return GADAdSizeLeaderboard;
   }
-  NSError *bannerError = nil;
-  [VungleSDK.sharedSDK addAdViewToView:_bannerView
-                           withOptions:options
-                           placementID:self.desiredPlacement
-                              adMarkup:[self bidResponse]
-                                 error:&bannerError];
-  return bannerError;
+  return GADAdSizeInvalid;
 }
 
-- (void)cleanUp {
-  if (_didBannerFinishPresenting) {
-    return;
+- (BannerSize)convertGADAdSizeToBannerSize {
+  if (GADAdSizeEqualToSize(_bannerSize, GADAdSizeMediumRectangle)) {
+    return BannerSizeMrec;
   }
-  _didBannerFinishPresenting = YES;
-
-  [VungleSDK.sharedSDK finishDisplayingAd:self.desiredPlacement adMarkup:self.bidResponse];
-  [[GADMAdapterVungleBiddingRouter sharedInstance] removeDelegate:self];
-  _bannerView = nil;
-}
-
-#pragma mark - GADMAdapterVungleDelegate delegates
-
-- (NSString *)bidResponse {
-  return [_adConfiguration bidResponse];
-}
-
-- (GADAdSize)bannerAdSize {
-  return _bannerSize;
-}
-
-- (void)initialized:(BOOL)isSuccess error:(nullable NSError *)error {
-  if (!isSuccess) {
-    _adLoadCompletionHandler(nil, error);
-    return;
+  if (_bannerSize.size.height == GADAdSizeLeaderboard.size.height) {
+    return BannerSizeLeaderboard;
   }
-  [self loadAd];
+  // Height is 50.
+  if (_bannerSize.size.width < GADAdSizeBanner.size.width) {
+    return BannerSizeShort;
+  }
+  return BannerSizeRegular;
 }
 
-- (void)adAvailable {
-  if (_isAdLoaded) {
-    // Already invoked an ad load callback.
-    return;
-  }
-  _isAdLoaded = YES;
+#pragma mark - VungleBannerDelegate
+
+- (void)bannerAdDidLoad:(VungleBanner *)banner {
   _bannerView = [[UIView alloc]
       initWithFrame:CGRectMake(0, 0, _bannerSize.size.width, _bannerSize.size.height)];
-  self.bannerState = BannerRouterDelegateStateWillPlay;
-  NSError *error = [self renderAd];
-  if (error) {
-    _adLoadCompletionHandler(nil, error);
-    return;
-  }
-
   if (_adLoadCompletionHandler) {
     _delegate = _adLoadCompletionHandler(self, nil);
+    [_bannerAd presentOn:_bannerView];
   }
+}
 
-  if (!_delegate) {
-    [[GADMAdapterVungleBiddingRouter sharedInstance] removeDelegate:self];
-    return;
-  }
+- (void)bannerAdDidFailToLoad:(VungleBanner *)banner withError:(NSError *)error {
+  NSError *gadError = GADMAdapterVungleErrorToGADError(GADMAdapterVungleErrorAdNotPlayable,
```

Comment:
The errors from Vungle SDK has been updated to have more precise error codes. So it will have a lot more values than the ones listed on the doc and newer codes may be made in the future.

So if this is removed, then the [documentation here](https://developers.google.com/admob/ios/mediation/vungle#error_codes) would need to be removed or updated to the latest. I think having the adapter map it to an existing code would be simpler on both our ends.